ga4gh/datamodel/sequenceAnnotations.py

```diff
@@ -0,0 +1,415 @@
+               "reference_name = ? "
+               "AND end > ? "  # compare this to query start
+               "AND start < ? "  # and this to query end
+               )
+        # TODO: Optimize by refactoring out string concatenation
+        sql_args = (referenceName, start, end)
+        if parentId is not None:
+            sql += "AND parent_id = ? "
+            sql_args += (parentId,)
+        if featureTypes is not None and len(featureTypes) > 0:
+            sql += "AND ontology_term IN ("
+            sql += ", ".join(["?", ] * len(featureTypes))
+            sql += ") "
+            sql_args += tuple(featureTypes)
+        query = self._dbconn.execute(sql, sql_args)
+        return (query.fetchone())[0]
+
+    def searchFeaturesInDb(self, pageToken=0, pageSize=None,
+                           referenceName=None, start=0, end=0,
+                           parentId=None, featureTypes=[]):
+        """
+        :param pageToken: int representing first record to return
+        :param pageSize: int representing number of records to return
+        :param referenceName: string representing reference name, ex 'chr1'
+        :param start: int position on reference to start search
+        :param end: int position on reference to end search >= start
+        :param parentId: string restrict search by id of parent node.
+        :return an array of dictionaries, representing the returned data.
+        """
+        # TODO: Refactor out common bits of this and the above count query.
+        sql = ("SELECT * FROM FEATURE WHERE "
+               "reference_name = ? "
+               "AND end > ? "  # compare this to query start
+               "AND start < ? "  # and this to query end
+               )
+        # TODO: Optimize by refactoring out string concatenation
+        sql_args = (referenceName, start, end)
+        if parentId is not None:
+            sql += "AND parent_id = ? "
+            sql_args += (parentId,)
+        if featureTypes is not None and len(featureTypes) > 0:
+            sql += "AND ontology_term IN ("
+            sql += ", ".join(["?", ] * len(featureTypes))
+            sql += ") "
+            sql_args += tuple(featureTypes)
+        sql += sqliteBackend.limitsSql(pageToken, pageSize)
+        query = self._dbconn.execute(sql, sql_args)
+        return sqliteBackend.sqliteRows2dicts(query.fetchall())
+
+    def getFeatureById(self, featureId):
+        """
+        :param featureId: the FeatureID as found in GFF3 records
+        :return: dictionary representing a feature object,
+            or None if no match is found.
+        """
+        sql = ("SELECT * FROM FEATURE WHERE id = ?")
+        query = self._dbconn.execute(sql, (featureId,))
+        ret = query.fetchone()
+        if ret is None:
+            return None
+        return sqliteBackend.sqliteRow2Dict(ret)
+
+
+# Problem: FeatureIds from GFF3 often contain colons,
+# and the CompoundId mechanism reserves colons for its special separator
```

Comment:
If we re-consider the colon, I'd like to explore using some mechanism other than non-alphanumeric characters, as the "native" ID's for features permit (and customarily use) a lot of wacky characters.